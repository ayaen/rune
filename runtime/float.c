//  Copyright 2024 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "runtime.h"
#include <stdint.h>
#include <stdio.h>
#include <float.h>
#include <string.h>

#define MAX_FLOAT_STRING_SIZE (64)

#if defined(DBL_DECIMAL_DIG)
#  define DBL_SIG_DIGITS (DBL_DECIMAL_DIG)
#  define FLT_SIG_DIGITS (FLT_DECIMAL_DIG)
#elif defined(DECIMAL_DIG)
#  define DBL_SIG_DIGITS (DECIMAL_DIG)
#  define FLT_SIG_DIGITS (DECIMAL_DIG)
#else
#  define DBL_SIG_DIGITS (DBL_DIG + 3)
#  define FLT_SIG_DIGITS (FLT_DIG + 3)
#endif


void runtime_f32tostring(runtime_array *dest, float value) {
  char tmp[MAX_FLOAT_STRING_SIZE];
  memset(tmp, 0, sizeof(tmp));
  int len = snprintf(tmp, sizeof(tmp), "%.*g", FLT_SIG_DIGITS, value);

  runtime_freeArray(dest);
  runtime_allocArray(dest, len, sizeof(uint8_t), false);
  memcpy((char*)dest->data, tmp, len);
}

void runtime_f64tostring(runtime_array *dest, double value) {
  char tmp[MAX_FLOAT_STRING_SIZE];
  memset(tmp, 0, sizeof(tmp));
  int len = snprintf(tmp, sizeof(tmp), "%.*g", DBL_SIG_DIGITS, value);

  runtime_freeArray(dest);
  runtime_allocArray(dest, len, sizeof(uint8_t), false);
  memcpy((char*)dest->data, tmp, len);
}
