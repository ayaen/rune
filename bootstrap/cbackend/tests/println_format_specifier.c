#include <stdio.h>
#include <float.h>
#include <inttypes.h>

#define STRINGIFY(s) STR(s)
#define STR(s) #s

#if defined(DBL_DECIMAL_DIG)
#  define DBL_SIG_DIGITS STRINGIFY(DBL_DECIMAL_DIG)
#  define FLT_SIG_DIGITS STRINGIFY(FLT_DECIMAL_DIG)
#elif defined(DECIMAL_DIG)
#  define DBL_SIG_DIGITS STRINGIFY(DECIMAL_DIG)
#  define FLT_SIG_DIGITS STRINGIFY(DECIMAL_DIG)
#else
#  define DBL_SIG_DIGITS STRINGIFY(DBL_DIG + 3)
#  define FLT_SIG_DIGITS STRINGIFY(FLT_DIG + 3)
#endif


void main(int argc, const char **argv) {
  printf("This is a string: \"%s\"\n", "hello");
  printf("This is an int: %" PRId8 "\n", (int8_t)(-1));
  printf("This is an unsigned int: %" PRIu16 "\n", (uint16_t)(65535u));
  printf("This is a hex value: 0x%" PRIx32 "\n", (uint32_t)(3735928559ul));
  printf("This is a 64-bit int: %" PRIu64 "\n", 1ull);
  printf("This is a 64-bit signed int: %" PRId64 "\n", -1ull);
  printf("This is a float: %." FLT_SIG_DIGITS "g\n", 2.7182817500000001);
  printf("This is a negative float: %." DBL_SIG_DIGITS "g\n", -3.1415926535897931);
}

