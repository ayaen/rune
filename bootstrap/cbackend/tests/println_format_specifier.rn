//  Copyright 2024 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Rune format specifiers:
//
//   %b        - Match an bool value: prints true or false
//   %i<width> - Match an Int value
//   %u<width> - Match a Uint value
//   %f        - Match a Float value
//   %s - match a string value
//   %x<width> - Match an Int or Uint value, print in lower-case-hex
//   %[t]      - Match an array of type t value, were t is another format specifier
//   %(t1, ..., tn) - Tuple value of format specifiers

println "This is a string: \"%s\"" % "hello"
println "This is an int: %i8" % -1i8
println "This is an unsigned int: %u16" % 65535u16
println "This is a hex value: 0x%x32" % 0xdeadbeefu32
println "This is a 64-bit int: %u64" % 1
println "This is a 64-bit signed int: %i64" % -1

println "This is a float: %f32" % 2.71828175f32
println "This is a negative float: %f64" % -3.1415926535897931

