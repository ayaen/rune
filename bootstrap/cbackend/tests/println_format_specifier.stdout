This is a string: "hello"
This is an int: -1
This is an unsigned int: 65535
This is a hex value: 0xdeadbeef
This is a 64-bit int: 1
This is a 64-bit signed int: -1
This is a float: 2.71828175
This is a negative float: -3.1415926535897931
